ext/BijectorsEnzymeCoreExt.jl

```diff
@@ -0,0 +1,184 @@
+    if !(
+        RT <: Union{Const,Duplicated,DuplicatedNoNeed} &&
+        wt_y isa Union{Const,Duplicated} &&
+        wt_u_hat isa Union{Const,Duplicated} &&
+        b isa Union{Const,Duplicated}
+    ) && !(
+        RT <: Union{Const,BatchDuplicated,BatchDuplicatedNoNeed} &&
+        wt_y isa Union{Const,BatchDuplicated} &&
+        wt_u_hat isa Union{Const,BatchDuplicated} &&
+        b isa Union{Const,BatchDuplicated}
+    )
+        throw(ArgumentError("inconsistent activities"))
+    end
+
+    # Compute primal value
+    Ω = find_alpha(wt_y.val, wt_u_hat.val, b.val)
+
+    # Early exit if no derivatives are requested
+    if RT <: Const
+        return Ω
+    end
+
+    Ω̇ = if wt_y isa Const && wt_u_hat isa Const && b isa Const
+        # Trivial case: All partial derivatives are 0
+        zero(Ω)
+    else
+        # In all other cases we have to compute the partial derivatives
+        ∂Ω_∂wt_y, ∂Ω_∂wt_u_hat, ∂Ω_∂b = ∂find_alpha(Ω, wt_y, wt_u_hat, b)
+        _muladd_partial(
+            ∂Ω_∂wt_y,
+            wt_y,
+            _muladd_partial(∂Ω_∂wt_u_hat, wt_u_hat, _muladd_partial(∂Ω_∂b, b, nothing)),
+        )
+    end
+
+    if RT <: Duplicated
+        @assert Ω̇ isa Real
+        return Duplicated(Ω, Ω̇)
+    elseif RT <: DuplicatedNoNeed
+        @assert Ω̇ isa Real
+        return Ω̇
+    elseif RT <: BatchDuplicated
+        @assert Ω̇ isa Tuple{Vararg{Real}}
+        return BatchDuplicated(Ω, Ω̇)
+    else
+        @assert RT <: BatchDuplicatedNoNeed
+        @assert Ω̇ isa Tuple{Vararg{Real}}
+        return Ω̇
+    end
+end
+
+function EnzymeRules.augmented_primal(
+    config::EnzymeRules.Config,
+    ::Const{typeof(find_alpha)},
+    ::Type{RT},
+    wt_y::Union{Const,Active},
+    wt_u_hat::Union{Const,Active},
+    b::Union{Const,Active},
+) where {RT<:Union{Const,Active}}
+    # Only compute the the original return value if it is actually needed
+    Ω =
+        if EnzymeRules.needs_primal(config) ||
+            EnzymeRules.needs_shadow(config) ||
+            (RT <: Active && !(wt_y isa Const && wt_u_hat isa Const && b isa Const))
+            find_alpha(wt_y.val, wt_u_hat.val, b.val)
+        else
+            nothing
+        end
+
+    tape = if RT <: Const || (wt_y isa Const && wt_u_hat isa Const && b isa Const)
+        # Trivial case: No differentiation or all partial derivatives are 0
+        # Thus no tape is needed
+        nothing
+    else
+        # Store the partial derivatives
+        ∂find_alpha(Ω, wt_y, wt_u_hat, b)
+    end
+
+    # Ensure that we follow the interface requirements of `augmented_primal`
+    primal = EnzymeRules.needs_primal(config) ? Ω : nothing
+    shadow = EnzymeRules.needs_shadow(config) ? zero(Ω) : nothing
+
+    return EnzymeRules.AugmentedReturn(primal, shadow, tape)
+end
+
+function EnzymeRules.reverse(
+    config::EnzymeRules.ConfigWidth{1},
+    ::Const{typeof(find_alpha)},
+    ::Type{<:Const},
+    ::Nothing,
+    wt_y::Union{Const,Active},
```

Comment:
at this point why bother having an annotation at all here?

Reply:
I want to ensure when and with which types of arguments the function is called, to be sure that it is actually doing the correct thing and does not give silently wrong results. Additionally, IMO it makes it easier to read and understand the code (non-annotated arguments and very generic types were quite confusing to me when I tried to read and learn from rules in Enzyme).